improver/regrid/bilinear.py

```diff
@@ -0,0 +1,502 @@
+            updated weights: source grid point weighting for all target grid points
+            updated surface_type_mask: - matching info between source/target point types
+    """
+
+    # increase 4 points to 8 points
+    out_latlons_updates = out_latlons[lake_island_indexes]
+
+    # Consider a larger area of 8 nearest points to look for more distant same
+    # surface type input points
+    # more than 8 points are within searching limits not considered here
+    k_nearest = 8
+    distances_updates, indexes_updates = nearest_input_pts(
+        in_latlons, out_latlons_updates, k_nearest
+    )
+
+    # Update output surface classification and surface type mask
+    out_classified_updates = out_classified[lake_island_indexes]
+    surface_type_mask_updates = similar_surface_classify(
+        in_classified, out_classified_updates, indexes_updates
+    )
+
+    # Where distance is outside specified vicinity, set surface type to be mismatched
+    # so that it will not be used, update surface type mask again
+    distance_not_in_vicinity = distances_updates > vicinity
+    surface_type_mask_updates = np.where(
+        distance_not_in_vicinity, False, surface_type_mask_updates
+    )
+
+    count_matching_surface = np.count_nonzero(surface_type_mask_updates, axis=1)
+    points_with_no_match = (np.where(count_matching_surface == 0))[0]
+
+    # If the expanded search area hasn't found any same surface type matches anywhere
+    # just ignore surface type, use normal bilinear approach
+    if points_with_no_match.shape[0] > 0:
+
+        # revert back to the basic bilinear weights, indexes unchanged
+        no_match_indexes = lake_island_indexes[points_with_no_match]
+        weights[no_match_indexes] = basic_weights(
+            no_match_indexes, indexes, out_latlons, in_latlons, in_lons_dim
+        )
+
+    points_with_match = (np.where(count_matching_surface > 0))[0]
+    # pylint: disable=unsubscriptable-object
+    count_of_points_with_match = points_with_match.shape[0]
+
+    # if no further processing can be done, return early
+    if count_of_points_with_match == 0:
+        return weights, indexes, surface_type_mask
+
+    # Where a same surface type match has been found among the 8 nearest inputs, apply
+    # inverse distance weighting with those matched points
+    new_distances = np.zeros([count_of_points_with_match, 4])
+    # pylint: disable=unsubscriptable-object
+    for point_idx in range(points_with_match.shape[0]):
+        match_indexes = lake_island_indexes[points_with_match[point_idx]]
+        # Reset all input weight and surface type to mismatched
+        weights[match_indexes, :] = 0.0
+        surface_type_mask[match_indexes, :] = False
+        # Loop through 8 nearest points found
+        good_count = 0
+        for i in range(k_nearest):
+            # Look for an input point with same surface type as output
+            if surface_type_mask_updates[points_with_match[point_idx], i]:
```

Comment:
only new matching points has surface_type_mask "True", otherwise "False". If it is mixed with old point, it does not matter as old surface mask is "False" (not considered in follow-up calculation. 
In equal distance cases, randomly picked ... based on list order in KDtree result.